lib/dao.js

```diff
@@ -2421,6 +2421,108 @@ DataAccessObject.updateAll = function(where, data, options, cb) {
   return cb.promise;
 };
 
+DataAccessObject.patchById = function patchById(id, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  assert(arguments.length >= 3, 'At least three arguments are required');
+
+  var Model = this;
+  var where = {};
+  var idPropertyName = idName(Model);
+
+  where[idPropertyName] = id;
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      // patchById(id, data, cb)
+      cb = options;
+      options = {};
+    }
+  }
+
+  data = data || {};
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+    'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = Model.getDataSource().connector;
+  assert(typeof connector.update === 'function',
+    'update() must be implemented by the connector');
+
+  var hookState = {};
+  this.applyProperties(data);
+
+  var context = {
+    Model: Model,
+    instanceId: id,
+    data: data,
+    hookState: hookState,
+    options: options,
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
+    doUpdate(ctx.instanceId, ctx.data);
```

Comment:
When we add `ctx.where` instead of `ctx.instanceId`, we need to check what is the `where` filter after hooks have run. See how `updateOrCreate` handles this situation - https://github.com/strongloop/loopback-datasource-juggler/blob/7247b6637e098047daf8d73d4d864455d80a9566/lib/dao.js#L493.

I am proposing to work incrementally and implement only the case when `ctx.where` stays pointing to `instanceId`. If `ctx.where` is changed, then `patchById` should fail with an error. Both branches should be covered with unit-tests though.

@Amir-61 Thoughts?